### DISCUSSION TITLE
How to utilize Catch All, Empty Placeholders, Priority routing features correctly since v23+?

### DISCUSSION THREAD

**Discussion by suchmememanyskill** [ThreeMammals/Ocelot#2136] [Q&A] [answered]
## Bug

Since release v23+, the [Catch All](https://ocelot.readthedocs.io/en/latest/features/routing.html#catch-all) functionality seems to be broken on sub-paths. Absolute routes no longer match over catch-all ones.

Screenshot 1 works as intended, Screenshot 2 does not.
This seems to be related to this PR: https://github.com/ThreeMammals/Ocelot/pull/1911

Expected behavior (Screenshot 1):
- `http://localhost:5000/` &rarr; `http://localhost:8000/specific/`
- `http://localhost:5000/url` &rarr; `http://localhost:8000/generic/url`

Actual behavior (Screenshot 1):
- `http://localhost:5000/` &rarr; `http://localhost:8000/specific/`
- `http://localhost:5000/url` &rarr; `http://localhost:8000/generic/url`

Screenshot 1:
![afbeelding](https://github.com/user-attachments/assets/0fceb360-d134-418b-b0cc-a08dacc510f6)

Expected behavior (Screenshot 2):
- `http://localhost:5000/test` &rarr; `http://localhost:8000/specific/`
- `http://localhost:5000/test/` &rarr; `http://localhost:8000/specific/`
- `http://localhost:5000/test/abc` &rarr; `http://localhost:8000/generic/abc`

Actual behavior (Screenshot 2):
- `http://localhost:5000/test` &rarr; `http://localhost:8000/generic/` (Unexpected!)
- `http://localhost:5000/test/` &rarr; `http://localhost:8000/generic/` (Unexpected!)
- `http://localhost:5000/test/abc` &rarr; `http://localhost:8000/generic/abc`

Screenshot 2:
![afbeelding](https://github.com/user-attachments/assets/7cec6768-5350-4f91-a20e-a94f0a3dead1)

## Specifications

  - Version: Ocelot 23.x
  - Platform: Windows, Dotnet 8.0.302
  - Subsystem: Kestrel

**Comment by suchmememanyskill** [comment]
This seems to also happen with multiple levels of catch all

Ocelot:
```json
  "Routes": [
    {
      "UpstreamHttpMethod": [ "Get" ],
      "UpstreamPathTemplate": "/test/{url}/{url2}",
      "DownstreamPathTemplate": "/generic/{url}/{url2}",
      "DownstreamScheme": "http",
      "DownstreamHostAndPorts": [
        {
          "Host": "localhost",
          "Port": 8000
        }
      ]
    },
    {
      "UpstreamHttpMethod": [ "Get" ],
      "UpstreamPathTemplate": "/test/{url}",
      "DownstreamPathTemplate": "/specific/{url}",
      "DownstreamScheme": "http",
      "DownstreamHostAndPorts": [
        {
          "Host": "localhost",
          "Port": 8000
        }
      ]
    }
  ],
  "GlobalConfiguration": {
    "BaseUrl": "https://localhost:5000"
  }
  ```
  
  Only `curl http://localhost:5000/test > nul` matches the /specific route. `http://localhost:5000/test/abc/a`, `http://localhost:5000/test/abc`, `http://localhost:5000/test/a` all match the /generic route.

**Comment by raman-m** [comment]
Hello Sims!
I see you like to have a testing fun. OK.

> Since release v23, the [catch all](https://ocelot.readthedocs.io/en/latest/features/routing.html#catch-all) functionality seems to be broken on subpaths. Absolute routes no longer match over catch-all ones.

Actually, the functionality was "broken"/changed specifically in version [23.0.0](https://github.com/ThreeMammals/Ocelot/releases/tag/23.0.0) as seen in the commit https://github.com/ThreeMammals/Ocelot/commit/f4803c24bf9e9ca3929c78ca8eb23401e3c31c23.

> Screenshot 1 works as intended, Screenshot 2 does not.

Understood. However, it's important to note that the two screenshots pertain to distinct features. Additional details follow.

> This seems to be related to this PR: https://github.com/ThreeMammals/Ocelot/pull/1911

Correct! It is connected to PR #1911. While #748 was initially closed with the bug fix in #1911, it turns out that #748 contains new functionality requirements, making #1911 a feature PR rather than a simple bug fix.

**Comment by raman-m** [comment]
> Expected behavior (Screenshot 1):
> - `http://localhost:5000/` &rarr; `http://localhost:8000/specific/`
> - `http://localhost:5000/url` &rarr; `http://localhost:8000/generic/url`
> 
> Actual behavior (Screenshot 1):
> - `http://localhost:5000/` &rarr; `http://localhost:8000/specific/`
> - `http://localhost:5000/url` &rarr; `http://localhost:8000/generic/url`
> 
> Screenshot 1:
> ![afbeelding](https://github.com/user-attachments/assets/0fceb360-d134-418b-b0cc-a08dacc510f6)
> 

The screenshot and user scenario depicted precisely illustrate the [Catch All](https://ocelot.readthedocs.io/en/latest/features/routing.html#catch-all) Routing feature. It is operational and continues to function in version 23.3. Perfect!

**Comment by suchmememanyskill** [comment]
Hi @raman-m,
Thank you for your quick reply

I'm a little confused by your response. Is this a bug or am i not using the features of Ocelot correctly?

> Understood. However, it's important to note that the two screenshots pertain to distinct features. Additional details follow.

So is the catch-all feature only supposed to be used for the root path? Can you link me the documentation for a similar feature for a catch-all on subpaths?

Also, am i correct to think that the input url that matches the upstream template more closely should be taken, rather than a priority from first to last? (Otherwise you'd be unable to create more advanced routing definitions)

Another question, can i disable the feature that's causing this as a temporary workaround?
(In my opinion it's a little absurd to add routing definitions that are not explicitly defined. How can i define very specific routes then that don't result in more paths than intended being mapped?)

**Reply by raman-m** [reply]
If you have read the entire topic, do you still require detailed answers to all those questions?
Anyway, my final answer is https://github.com/ThreeMammals/Ocelot/discussions/2136#discussioncomment-10223003 ❗

**Reply by suchmememanyskill** [reply]
I wrote this at the same time as you were writing part of the response. You can safely disregard this post

**Comment by raman-m** [comment]
> **Expected behavior** (Screenshot 2):
> - `http://localhost:5000/test` &rarr; `http://localhost:8000/specific/`
> - `http://localhost:5000/test/` &rarr; `http://localhost:8000/specific/`
> - `http://localhost:5000/test/abc` &rarr; `http://localhost:8000/generic/abc`
> 
> **Actual behavior** (Screenshot 2):
> - `http://localhost:5000/test` &rarr; `http://localhost:8000/generic/` (Unexpected!)
> - `http://localhost:5000/test/` &rarr; `http://localhost:8000/generic/` (Unexpected!)
> - `http://localhost:5000/test/abc` &rarr; `http://localhost:8000/generic/abc`
> 
> Screenshot 2:
> ![afbeelding](https://github.com/user-attachments/assets/7cec6768-5350-4f91-a20e-a94f0a3dead1)
> 

The "Screenshot 2" user scenario pertains to the [Empty Placeholders](https://ocelot.readthedocs.io/en/latest/features/routing.html#empty-placeholders) feature. Both upstream templates `/test/{url}` and `/test/` are valid for the [Empty Placeholders](https://ocelot.readthedocs.io/en/latest/features/routing.html#empty-placeholders) feature, making both routes applicable. In line with the logic for applicable routes and the [Priority](https://ocelot.readthedocs.io/en/latest/features/routing.html#priority) routing feature, Ocelot's internal logic selects all suitable routes and assigns priorities as shown in this code snippet:
https://github.com/ThreeMammals/Ocelot/blob/b4e21c47106acc8577862203356c3e33280a2985/src/Ocelot/DownstreamRouteFinder/Finder/DownstreamRouteFinder.cs#L32-L34
Subsequently, the `foreach` loop iterates over all the applicable routes and generates a new list of downstream routes:
https://github.com/ThreeMammals/Ocelot/blob/b4e21c47106acc8577862203356c3e33280a2985/src/Ocelot/DownstreamRouteFinder/Finder/DownstreamRouteFinder.cs#L36
However, since the Priority feature is not employed, the first route in the list is selected by default.

Additionally, it can be said that the **Actual behavior** is derived from the [Empty Placeholders](https://ocelot.readthedocs.io/en/latest/features/routing.html#empty-placeholders) feature. Conversely, the **Expected behavior** is a result of the [Priority](https://ocelot.readthedocs.io/en/latest/features/routing.html#priority) feature.

### To achieve the expected behavior, one must employ the [Priority](https://ocelot.readthedocs.io/en/latest/features/routing.html#priority) routing feature, which is specifically designed for scenarios with implicit or unclear route priorities.

**Comment by suchmememanyskill** [comment]
> To achieve the expected behavior, one must employ the [Priority](https://ocelot.readthedocs.io/en/latest/features/routing.html#priority) routing feature, which is specifically designed for scenarios with implicit or unclear route priorities.

For https://github.com/ThreeMammals/Ocelot/discussions/2136#discussioncomment-10222996, it's impossible to use priority to solve this. The first match will always swallow the entire url.

**Comment by raman-m** [comment]
@suchmememanyskill [wrote](https://github.com/ThreeMammals/Ocelot/discussions/2136#discussioncomment-10222996)
> This seems to also happen with multiple levels of catch all

I understand your interest in software testing, but what is the purpose of such detailed analysis? Both `"/test/{url}/{url2}"` and `/test/{url}` are logically equivalent. What is the reason for this discussion? Theoretically, you might want to extract the first `{url}` subpath and apply it in custom C# logic, but is there actually any custom logic in place?

>   Only `curl http://localhost:5000/test > nul` matches the /specific route.

Indeed, the URL matches the second route template because it does not have two subpaths required to match the first route. Bingo!

> `http://localhost:5000/test/abc/a`, `http://localhost:5000/test/abc`, `http://localhost:5000/test/a` all match the /generic route.

Indeed, the URLs `http://localhost:5000/test/abc/a`, `http://localhost:5000/test/abc`, and `http://localhost:5000/test/a` all correspond to the /generic route. This is due to each URL containing at least one subpath, which triggers the first route to be applicable under the internal criteria of the [Empty Placeholders](https://ocelot.readthedocs.io/en/latest/features/routing.html#empty-placeholders) feature.

All behavior is correct! There are no issues. However, there seems to be a lack of clarity regarding the [Routing](https://ocelot.readthedocs.io/en/latest/features/routing.html) feature, which warrants a second reading by you. Experimenting with subpaths while disregarding the existing, ready-to-use features is nonsensical. Such experiments often lead to confusion and are prone to failure. It's advisable to fully utilize the power of the already implemented [Routing](https://ocelot.readthedocs.io/en/latest/features/routing.html) features before attempting to develop more customized solutions on top of the existing functionality.

**Reply by suchmememanyskill** [reply]
@raman-m  Fair point. I was testing the limits of the routing engine. A potential usecase is routing /{a}/{b} to a different server than /{a}. (One service fetches lists, one fetches specific data).

Either way, thank you for you explanations.
The behaviour seems to be intended.

One final question, then.

Can the [empty placeholders](https://ocelot.readthedocs.io/en/latest/features/routing.html#empty-placeholders) feature be disabled? I'd prefer when i define a specific route, (like /test/{data}), that /test/ and /test isn't routed to my server. This for example can be a bad idea in combination with medical standards like [fhir](http://hl7.org/fhir/STU3/search.html). Especially if different servers handle the specific incoming requests

**Reply by raman-m** [reply]
> Can the [empty placeholders](https://ocelot.readthedocs.io/en/latest/features/routing.html#empty-placeholders) feature be disabled?

Certainly, you can define an alternative upstream template path prefix such as `/specific/{data}` instead of `/test/{data}`, while still directing to the identical downstream service.

**Reply by suchmememanyskill** [reply]
@raman-m I think you misread my question.

i want just `/upstream/{data}` to be routed to `/downstream/{data}`. I don't want `/upstream/` and `/upstream` to be routed to `/downstream`. Not being able to do this will make routing fhir data unnecessarily difficult.

![afbeelding](https://github.com/user-attachments/assets/dfd3508e-2f31-4832-b236-80620da3e2c0)
Specifically, the last 2 incoming request are unwanted routes.

If i downgrade to v22, this behaviour works as i would personally expect it to work. Downgrading to me is not an option, though (do we really want to put people on unmaintained versions just to have different routing behaviours?).
![afbeelding](https://github.com/user-attachments/assets/59d81b17-5474-4174-bbe2-6a13336e02a2)

**Reply by raman-m** [reply]
> I don't want `/upstream/` and `/upstream` to be routed to `/downstream`

There are two potential approaches to continue: 
- Remain indefinitely on version 22.x.
- Upgrade to version 23.x or higher and designate high-priority routes to restrict them.

Currently, Ocelot does not have a feature to explicitly forbid URLs. However, you can achieve this by using a fake URL in the downstream host and port settings (such as a non-existent localhost port) to prevent redirection to a real endpoint. This will result in errors in your logs, but it will effectively block specific URLs by catching and forbidding them in `ocelot.json`.

Alternatively, you could create a specialized path filtering [Delegating Handler](https://ocelot.readthedocs.io/en/latest/features/delegatinghandlers.html) or [Overriding Middleware](https://ocelot.readthedocs.io/en/latest/features/middlewareinjection.html) with paths defined by [Metadata](https://ocelot.readthedocs.io/en/latest/features/metadata.html). While this method necessitates C# development, it may result in error-free logs if executed correctly.

**Comment by raman-m** [comment, marked as answer]
@suchmememanyskill [wrote](https://github.com/ThreeMammals/Ocelot/discussions/2136#discussioncomment-10223001)
> For https://github.com/ThreeMammals/Ocelot/discussions/2136#discussioncomment-10222996, it's impossible to use priority to solve this. The first match will always swallow the entire url. 

I haven't had the time to verify the accuracy of the statement, but it appears everything is detailed in my https://github.com/ThreeMammals/Ocelot/discussions/2136#discussioncomment-10223002.
Priorities are only sensible for applicable routes! In this instance, since the routes differ and each URL has just one applicable route, the Priority feature becomes redundant. Correct? Thus, for this scenario, you would need to use a different upstream path prefix, and a possible solution could be as follows:

```json
  "Routes": [
    {
      "UpstreamPathTemplate": "/generic/{url}/{url2}",
      "DownstreamPathTemplate": "/generic/{url}/{url2}",
    },
    {
      "UpstreamPathTemplate": "/specific/{url}",
      "DownstreamPathTemplate": "/specific/{url}",
    }
  ],
  "GlobalConfiguration": {
    "BaseUrl": "https://localhost:5000"
  }
```
This solution is elegant and avoids the need for debates with the product owner. 😉
Best of luck with your endeavor with Ocelot!
